[RPD-2976] code refactoring

ehdudgnsTW opened the pull request:
<!--
  제목은 `[지라이슈번호] 작업 내용 요약`으로 작성해주세요.
-->

## 설명
현재 작업되어있던 코드들을 리팩토링 하였습니다.
### 작업 내용
기존에 MainViewController,FavoriteDetailViewController에서 의미 없이 뷰 만드는 작업을 삭제 하였습니다.
레파지토리를 싱글톤패턴으로 변경 하였습니다.
기존에 데이터모델들이 뷰컨트롤러로 직접 전달되는 방식을 리엑터를 통해서 전달되도록 변경 하였습니다.
테이블뷰셀 또한 셀리엑터를 생성하여서 셀리엑터를 통해서 데이터를 전달하도록 변경 하였습니다.
MainViewReactor와 DetailViewReactor에서 의미없는 작업을 제외하고 정리하여서 재구성 하였습니다.
### 스크린샷
<!--
  (Optional) UI가 변경되었다면 수정 전/후 이미지를 추가해주세요.
-->

## 기타
<!--
  (Optional)
  - 리뷰 시에 유심히 봐주었으면 하는 부분을 알려주세요.
  - merge 전 필요한 작업이 있다면 알려주세요.
  - 기타 등등 자유롭게 작성해주세요.
-->

## 참고 링크
<!--
  (Optional) 검토시 참고할 만한 자료를 공유해주세요.
  - 기획 문서, 디자인 문서, slack 메시지의 링크를 추가합니다.
  - 개발시 참고했던 기술 문서, article 등을 추가합니다.
-->

ehdudgnsTW committed:
Refactor Repository

repository 싱글톤 으로 변경하였습니다.
init할때 json파일 설정을 하도록 변경하였습니다.
이제 getData를 통해서 데이터를 받아옵니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewReactor

두개로 나누어져있던 search와 favorite을 한개로 합쳤습니다.
이제는 State를 통해서 검색중인지 아닌지를 알수 있습니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewController

기존에 bind하는 부분의 액션을 통일 시켰습니다.
이제는 isSearching이라는 변수를 삭제 하였습니다.
기존에 의미없는 view 만들어주는 부분을 삭제 하였습니다.

RPD-2976

ehdudgnsTW committed:
Add CellReactor

기존에 셀에서 모델을 받아오는 방식을 reactor를 통해서 받아오려고 파일 생성 하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor DetailViewReactor

모델을 뷰컨트롤러가 아닌 리엑터에서 받아오고 뷰컨트롤러로 넘기게 재구성 하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor FavoriteDetailDataViewController

모델을 받아와서 cofigure하는 작업을 모델이 아닌 리엑터에서 받아와서 bind부분에서 configure하도록 변경 하였습니다.
의미없는 view만드는 작업을 삭제하였습니다.

RPD-2976

ehdudgnsTW committed:
Fix TableViewCells

모델을 받아와서 configure하는 작업을 cellreactor를 통해서 데이터를 configure하도록 변경 하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewController

검색할때 기존에 검색이 남아있다면 그 검색 단어가 포함된 정보를 보여주도록 변경 하였습니다.
tableviewcell을 만들때 cellreactor를 통해서 정보를 넘겨주도록 변경하였습니다.
이제 상세페이지로 넘어갈때 모델을 직접 뷰컨트롤러에 넘겨주는 것이 아니라 상세페이지 리엑터에 넘겨주도록 수정 하였습니다.

RPD-2976

ehdudgnsTW committed:
Add action in MainViewReactor

셀에서 버튼을 누르는 액션을 추가하였습니다.
이때 검색어를 가지고 있어서 그 검색된 데이터를 보존하기 위해서 searchword라는 변수를 추가 하였습니다.

RPD-2976

ehdudgnsTW committed:
Fix data update issue

상세페이지에서 즐겨찾기 상태를 변경하고 나서 즐겨찾기 목록으로 돌아왔을때 즐겨찾기 목록이 업데이트 안되는 문제점을 해결 하였습니다.

RPD-2976

hyeonsikTW commented on TwinnyAssignmentiOSProject/ViewModels/CellReactor.swift:
```diff
@@ -0,0 +1,22 @@
+//
+//  CellReactor.swift
+//  TwinnyAssignmentiOSProject
+//
+//  Created by twinny on 2022/04/21.
+//
+
+import Foundation
+
+struct CellReactor {
+    var cityName: String
+    var cityId: String
+    var currentTemperature: String
+    var isFavorite: Bool
+    
+    init(model: FavoriteDataModel) {
+        self.cityName = model.cityName
+        self.cityId = model.cityId
+        self.currentTemperature = "\(model.currentTemperature)"
+        self.isFavorite = model.isFavorite
+    }
+}
```
CellReactor는 구조체여서 Models 폴더에 있는게 더 자연스러울 거 같습니다.

ehdudgnsTW committed:
Add keyboard dismiss

검색도중에 테이블뷰를 스크롤하면 키보드가 사라지게 만들었습니다.

RPD-2976

hyeonsikTW commented on TwinnyAssignmentiOSProject/ViewModels/MainViewReactor.swift:
```diff
@@ -13,84 +13,77 @@ import UIKit
+                else {
+                    self.searchWord = ""
+                    return repository.getData(nil)
+                        .map { $0 }
+                        .map { .filteringData($0, true) }
```
.map { $0 } 이거는 없어도 될 거 같습니당.

ehdudgnsTW committed:
Refactor Repository

repository파일을 싱글톤 디자인 패턴을 적용하여 변경 하였습니다.

RPD-2976

ehdudgnsTW committed:
Fix call repository in Reactors

리엑터에서 repository를 통해서 데이터를 불러오는 부분을 수정하였습니다.
또한 함수 하나가 하나의 작업을 하도록 변경 하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor CellReactor

기존의 cell reactor를 reactor 형식으로 변경 하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor LocationDataTableViewCell

기존에 configure로 정보를 받아오던 것들을 이제는 cellreactor를 통하여 받아오게 만들었습니다.

RPD-2976

ehdudgnsTW committed:
Refactor FavoriteDataTableViewCell

기존에 configure로 정보를 받아오던 것들을 이제는 cellreactor를 통하여 받아오게 만들었습니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewController

cell의 delegate를 추가하여 cell에서 버튼을 눌렀을때 delegate를 통해서 tableView를 reload 하게끔 만들었습니다.
cell을 선언할때 cell의 reactor를 초기화 한뒤에 cell.reactor에 넣어주었습니다.

RPD-2976

ehdudgnsTW committed:
Improve FavoriteDetailDataViewController

state에서 받아오는 부분을 조금더 효율적으로 변경 하였습니다.

RPD-2976

ehdudgnsTW committed:
Rename reloadTableView to updateFavoriteDatas

reloadTableView를 updateFavoriteDatas로 이름을 변경 하였습니다.

RPD-2976

ehdudgnsTW committed:
Fix temperature value change

즐겨찾기 버튼을 눌렀을때 값이 변하는 현상을 발견해서 이를 수정하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor CellReactor

cellreactor에 isSearching을 추가 해서 cellreactor를 통해서 서칭중인지 아닌지의 데이터를 가지게 만들었습니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewReactor

전체 데이터를 받아오는 부분이 수정되었습니다.
cellReactor의 배열로 반환하게 만들었습니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewController

기존에 bind하는 부분이 수정되었습니다.
서칭이 되던 중에 상세페이지로 넘어갔을때 이전 페이지로 돌아오게 구현했습니다.
단어 검색중에 cancel버튼을 눌렀을때 즐겨찾기 화면으로 안가는 이슈를 해결하였습니다.

RPD-2976

ehdudgnsTW committed:
Add delegate

아무값도 입력 하지 않았지만 검색중인 상태에서 상세페이지로 들어갔을때 뒤로가기를 했을때 즐겨찾기 화면으로 나오는 이슈를 delegate를 통해 해결 하였습니다.

RPD-2976

ehdudgnsTW committed:
Fix layout cell

기존에 경고가 뜨는 layout을 수정하였습니다.

RPD-2976

ehdudgnsTW committed:
Fix bind function

cell reactor의 state가 변경됨에 따라 값을 받아오는 부분을 수정 하였습니다.

RPD-2976

ehdudgnsTW committed:
Remove delegate in FavoriteDetailDataViewController

delegate를 통해서 isSearching을 받고 다시 보내주는 작업에 대해서 불필요한 작업이어서 삭제를 하였습니다.

RPD-2976

ehdudgnsTW committed:
Remove CellDataModel in CellReactor

cellReactor에서 불필요하게 감싸져있던 데이터 struct를 삭제하고 전역 변수들로 변경하였습니다.

RPD-2976

ehdudgnsTW committed:
Set ui to reactor inistualState

struct를 삭제하면서 기존에 매칭해주던 방식을 변경하였습니다.

RPD-2976

ehdudgnsTW committed:
Refactor MainViewController

cell Reactor의 변경으로 인해서 선언할때의 변경이 있었습니다.
불필요한 flag설정을 searchController.isActive로 변경 하였습니다.
DetailViewController로 넘기는 delegate를 삭제 하였습니다.

RPD-2976

ehdudgnsTW committed:
Improve layout LocationCell

기존에 있던 locationCell의 label layout을 변경 하였습니다.

RPD-2976

nexmondTW commented:
수고하셨습니다. 합칩니다.
